classes with user-written initializers allow param fields with default values that are not params.

<!--
If you are filing an issue that is not a bug report, please feel free to erase
this template and describe the issue as clearly as possible.
-->


### Summary of Problem

```chapel
config var x = 7;

class C {
  param p = x;
  proc init() { }
}

var c = new C();
writeln(c);
delete c;
```

Without the ``writeln(c)``, this compiles.  With it,

```
fortytwo@magrathea:~/src/chpl/tfn$ chpl --force-initializers wrparmem.chpl
In file included from /tmp/chpl-fortytwo-24984.deleteme/_main.c:28:0:
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c: In function ‘writeThisFieldsDefaultImpl2’:
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c:236:26: error: ‘tmp_chpl’ undeclared (first use in this function)
 writeIt_chpl(_ref_tmp_6, tmp_chpl);
                          ^
/tmp/chpl-fortytwo-24984.deleteme/ChapelIO.c:236:26: note: each undeclared identifier is reported only once for each function it appears in
/home/fortytwo/src/chapel/runtime/etc/Makefile.exe:27: recipe for target '/tmp/chpl-fortytwo-24984.deleteme/wrparmem.tmp' failed
make: *** [/tmp/chpl-fortytwo-24984.deleteme/wrparmem.tmp] Error 1
error: compiling generated source
```

Curiously, with the ``proc init() { }`` removed but with ``--force-initializers``, it reports

```
wrparmem.chpl:6: error: default value for param is not a param
```

which is what I expected to happen as written above.

**Associated Future Test(s):**

[`test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl)

[edit 2/12 cassella to add path to future]

### Configuration Information

```
fortytwo@magrathea:~/src/chpl/tfn$ chpl --version
chpl version 1.17.0 pre-release (44fa845)
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)

fortytwo@magrathea:~/src/chpl/tfn$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none

fortytwo@magrathea:~/src/chpl/tfn$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.6) 5.4.0 20160609
```

--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -4939,7 +4939,7 @@ static void resolveInitField(CallExpr* call) {
     // Check against the symbol's Immediate is required because the symbol is
     // likely a temporary, which generally will suppress an error if
     // initialized from a non-param.
-    if (srcSym->isParameter() == false ||
+    else if (srcSym->isParameter() == false ||
         (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
       USR_FATAL_CONT(call, "Initializing parameter '%s' to value not known at compile time", fs->name);
     }

--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -4926,7 +4926,8 @@ static void resolveInitField(CallExpr* call) {
   if (!fs)
     INT_FATAL(call, "bad initializer set field primitive");
 
-  Type* srcType = srcExpr->symbol()->type;
+  Symbol* srcSym = srcExpr->symbol();
+  Type* srcType = srcSym->type;
   if (srcType == dtUnknown)
     INT_FATAL(call, "Unable to resolve field type");
 
@@ -4934,7 +4935,12 @@ static void resolveInitField(CallExpr* call) {
     if (isLegalParamType(srcType) == false) {
       USR_FATAL_CONT(fs, "'%s' is not of a supported param type", fs->name);
     }
-    if (srcExpr->symbol()->isParameter() == false) {
+
+    // Check against the symbol's Immediate is required because the symbol is
+    // likely a temporary, which generally will suppress an error if
+    // initialized from a non-param.
+    if (srcSym->isParameter() == false ||
+        (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
       USR_FATAL_CONT(call, "Initializing parameter '%s' to value not known at compile time", fs->name);
     }
   }
@@ -4967,7 +4973,7 @@ static void resolveInitField(CallExpr* call) {
         (fs->defPoint->exprType == NULL && fs->defPoint->init == NULL) ||
         (fs->defPoint->init == NULL && fs->defPoint->exprType != NULL &&
          ct->fieldIsGeneric(fs, ignoredHasDefault))) {
-      AggregateType* instantiate = ct->getInstantiation(srcExpr->symbol(), index);
+      AggregateType* instantiate = ct->getInstantiation(srcSym, index);
       if (instantiate != ct) {
         // TODO: make this set of operations a helper function I can call
         INT_ASSERT(parentFn->_this);

--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -4939,6 +4939,10 @@ static void resolveInitField(CallExpr* call) {
     // Check against the symbol's Immediate is required because the symbol is
     // likely a temporary, which generally will suppress an error if
     // initialized from a non-param.
+    //
+    // Note: This unfortunately relies on not checking the error when
+    // initializing the temporary, where it would be more difficult to
+    // determine the context of the initialization.
     else if (srcSym->isParameter() == false ||
             (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
       USR_FATAL_CONT(call, "Initializing parameter '%s' to value not known at compile time", fs->name);

--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -4940,7 +4940,7 @@ static void resolveInitField(CallExpr* call) {
     // likely a temporary, which generally will suppress an error if
     // initialized from a non-param.
     else if (srcSym->isParameter() == false ||
-        (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
+            (isVarSymbol(srcSym) && getImmediate(srcSym) == NULL)) {
       USR_FATAL_CONT(call, "Initializing parameter '%s' to value not known at compile time", fs->name);
     }
   }

--- a/test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-config-var.chpl
@@ -4,6 +4,5 @@ class C {
   proc init() { }
 }
 
-var c = new C();
+var c = new owned C();
 writeln(c);
-delete c;

--- a/test/classes/initializers/generics/param-field-default-value-non-param-config-var.future
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-config-var.future
@@ -1,5 +1,0 @@
-bug: class with initializer allows non-param default value for param field
-#8423
-
-Without the writeln() it compiles.
-With the writeln(), there's an error compiling the generated ChapelIO.c.

--- a/test/classes/initializers/generics/param-field-default-value-non-param-config-var.good
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-config-var.good
@@ -1,1 +1,2 @@
-param-field-default-value-non-param-config-var.chpl:3: error: default value for param is not a param
+param-field-default-value-non-param-config-var.chpl:4: In initializer:
+param-field-default-value-non-param-config-var.chpl:4: error: Initializing parameter 'p' to value not known at compile time

--- a/test/classes/initializers/generics/param-field-default-value-non-param-field.chpl
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-field.chpl
@@ -4,6 +4,5 @@ class C {
   proc init() { }
 }
 
-var c = new C();
+var c = new owned C();
 writeln(c);
-delete c;

--- a/test/classes/initializers/generics/param-field-default-value-non-param-field.future
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-field.future
@@ -1,2 +1,0 @@
-bug: compiler segfault on class param field w/default value from non-param field
-#8443

--- a/test/classes/initializers/generics/param-field-default-value-non-param-field.good
+++ b/test/classes/initializers/generics/param-field-default-value-non-param-field.good
@@ -1,1 +1,2 @@
-param-field-default-value-non-param-field.chpl:3: error: default value for param is not a param
+param-field-default-value-non-param-field.chpl:4: In initializer:
+param-field-default-value-non-param-field.chpl:4: error: Initializing parameter 'p' to value not known at compile time

